Add Fine-tuning a Multimodal Model Using SFT (Single or Multi-Image Dataset) guide to docs (#3235)
huggingface/trl: docs/source/training_vlm_sft.md

        images = [process_vision_info(example["messages"]) for example in examples]

    # Tokenize the texts and process the images
    batch = processor(
        text=texts, images=images, return_tensors="pt", padding=True
    )  # Encode texts and images into tensors

    # The labels are the input_ids, and we mask the padding tokens in the loss computation
    labels = batch["input_ids"].clone()  # Clone input IDs for labels
    # Mask image tokens
    image_token_id = [
        processor.tokenizer.convert_tokens_to_ids(processor.tokenizer.special_tokens_map["boi_token"])
    ]
    # Mask tokens for not being used in the loss computation
    labels[labels == processor.tokenizer.pad_token_id] = -100
    labels[labels == image_token_id] = -100
    labels[labels == 262144] = -100

    batch["labels"] = labels
    return batch  # Return the prepared batch
```

### **Training the Model**  

With all the components set up, we now configure the `SFTTrainer` using the previously defined settings and start the training process.

``` python
# Training
from trl import SFTTrainer

trainer = SFTTrainer(
    model=model,
    args=training_args,
    data_collator=collate_fn,

>     train_dataset=dataset["train"], # multi-image -> train_dataset=dataset["test"],

Author: The dataset used for multi-image is a benchmark and it only comes with test set.

# Fine-tuning a Multimodal Model Using SFT (Single or Multi-Image Dataset)

> ![Single Image + Text](./training_vlm_sft_training_procedure_single_image.png)  

Reviewer: could you please use https://huggingface.co/datasets/trl-lib/documentation-images for the images? I've just sent you an invite to join the org

## Overview  

This guide walks you through the process of fine-tuning a multimodal language model (e.g., **Gemma 3**) using **Supervised Fine-Tuning (SFT)**. We cover two cases:  

- **Single Image + Text**  
- **Multi-Image + Text**  

This guide serves as a **detailed walkthrough** and complements the existing [VLM SFT script](https://github.com/huggingface/trl/blob/main/examples/scripts/sft_vlm_gemma3.py). If you're already familiar with the concepts, you can use the script directly.  

We demonstrate the fine-tuning process using two datasets, but these principles extend to other **Vision-Language Models (VLMs)** and datasets.  

## Understanding the Datasets  

To address both **Single Image + Text** and **Multi-Image + Text** scenarios, we use two datasets that are well-suited for this task.  

> ### **HuggingFaceH4/llava-instruct-mix-vsft Dataset (Image + Text)**  

Reviewer: ```suggestion
### HuggingFaceH4/llava-instruct-mix-vsft Dataset (Image + Text)
```

This dataset is a reformatted version of [LLaVA Instruct Mix](https://huggingface.co/datasets/theblackcat102/llava-instruct-mix). It consists of conversations where a user provides both **text** and a **single image** as input.  

The model (referred to as the **"assistant"**) responds based on both the **visual and textual information** shared by the user. This dataset is particularly useful for training multimodal models to **understand and generate responses based on images and text**.  

<iframe
  src="https://huggingface.co/datasets/HuggingFaceH4/llava-instruct-mix-vsft/embed/viewer/default/train"
  frameborder="0"
  width="100%"
  height="560px"
></iframe>

> ### **FanqingM/MMIU-Benchmark Dataset (Multi-Image + Text)**  

Reviewer: ```suggestion
### FanqingM/MMIU-Benchmark Dataset (Multi-Image + Text)
```

The **FanqingM/MMIU-Benchmark** dataset consists of:  

- **Context:** Included in the system prompt.  
- **Question:** Provided as part of the user's input.  
- **Series of Images:** Multiple images related to the question.  
- **Answer:** The model's expected response.  

This dataset is designed for tasks where the model must reason over multiple images to generate an informed response based on both visual and textual inputs.

<iframe
  src="https://huggingface.co/datasets/FanqingM/MMIU-Benchmark/embed/viewer/default/test"
  frameborder="0"
  width="100%"
  height="560px"
></iframe>

> ## **Developing a Fine-Tuning Script for Multimodal SFT**  

Reviewer: ```suggestion
## Developing a Fine-Tuning Script for Multimodal SFT
```

In this section, we build the script needed to fine-tune a multimodal model for both **Single Image + Text** and **Multi-Image + Text** use cases.  

> ### **Setting Up the Environment**  

Reviewer: ```suggestion
### Setting Up the Environment
```

Before fine-tuning, we need to install the required dependencies. Let's start by setting up the environment:  

```bash
# Install the required libraries

> pip install -U -q trl transformers datasets bitsandbytes peft accelerate hf_xet

Reviewer: ```suggestion
pip install -U -q trl bitsandbytes peft hf_xet
```

these are trl dependencies

# Alternative: Install a specific Transformers release with zero-day model support  
# pip install git+https://github.com/huggingface/transformers@v4.49.0-Gemma-3  

# Optional: Install the latest development version of TRL  

> # pip install git+https://github.com/huggingface/trl.git  

Reviewer: To simplify, you can refer to the installation section of the doc

    return dataset

dataset = prepare_dataset(dataset, dataset_name, dataset_train_split)
```

With this, your **Multi-Image + Text** dataset is now prepared for training.

### **Preparing for Training**  

We start by loading the model and processor. In this example, we use `google/gemma-3-4b-it`, but the same process applies to its other variants and similar models.  

To optimize memory usage, we configure `BitsAndBytes` to load the quantized version of the model.

```python
import torch
from transformers import AutoModelForImageTextToText, AutoProcessor, BitsAndBytesConfig

model_id = "google/gemma-3-4b-it"

# BitsAndBytesConfig int-4 config
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_quant_storage=torch.bfloat16,
)

# Load model and tokenizer
model = AutoModelForImageTextToText.from_pretrained(
    model_id, 
    device_map="auto", 
    torch_dtype=torch.bfloat16, 
    attn_implementation="eager", 
    quantization_config=bnb_config
)
processor = AutoProcessor.from_pretrained(model_id)
```

> Next, we set up [Quantized Low-Rank Adaptation (QLoRA)](https://arxiv.org/abs/2305.14314), an efficient fine-tuning technique for Large Language Models (LLMs) and Vision-Language Models (VLMs).  

Reviewer: ```suggestion
Next, we set up [Quantized Low-Rank Adaptation (QLoRA)](https://huggingface.co/papers/2305.14314), an efficient fine-tuning technique for Large Language Models (LLMs) and Vision-Language Models (VLMs).  
```

```python
from peft import LoraConfig, get_peft_model

# Configure QLoRA

peft_config = LoraConfig(
    lora_alpha=16,
    lora_dropout=0.05,
    r=16,
    bias="none",
    target_modules="all-linear",
    task_type="CAUSAL_LM",
    modules_to_save=[
        "lm_head",
        "embed_tokens",
    ],
)
```

> With QLoRA now set up, we need to define the training arguments for SFT. The `SFTConfig` class simplifies this process, providing an easy way to adjust parameters based on our specific needs.  

Reviewer: ```suggestion
With QLoRA now set up, we need to define the training arguments for SFT. The [`SFTConfig`] class simplifies this process, providing an easy way to adjust parameters based on our specific needs.  
```

```python
from trl import SFTConfig

training_args = SFTConfig(
    output_dir="gemma-3-4b-it-trl-sft-llava-instruct-mix-vsft",     # Directory to save the model and push to the Hub. Use a specific repository id (e.g., gemma-3-4b-it-trl-sft-MMIU-Benchmark for multi-image datasets).
    num_train_epochs=1,                                             # Set the number of epochs to train the model.
    per_device_train_batch_size=1,                                  # Batch size for each device (e.g., GPU) during training.

>     gradient_accumulation_steps=1,                                  # Number of steps before performing a backward/update pass to accumulate gradients.

Reviewer: is it enough, an effective batch size of 1? (assuming you use 1 GPU)

    gradient_checkpointing=True,                                    # Enable gradient checkpointing to reduce memory usage during training.
    optim="adamw_torch_fused",                                      # Use the fused AdamW optimizer for better performance.
    logging_steps=10,                                               # Frequency of logging training progress (log every 10 steps).
    save_strategy="epoch",                                          # Save checkpoints at the end of each epoch.
    learning_rate=2e-05,                                            # Learning rate for training.
    bf16=True,                                                      # Enable bfloat16 precision for training to save memory and speed up computations.
    push_to_hub=True,                                               # Automatically push the fine-tuned model to Hugging Face Hub after training.
    gradient_checkpointing_kwargs={"use_reentrant": False},         # Set gradient checkpointing to non-reentrant to avoid issues.

>     dataset_text_field="",                                          # Placeholder for dataset text field (needed by the collator).

Reviewer: ```suggestion
```

not needed anymore

    dataset_kwargs={"skip_prepare_dataset": True},                  # Skip dataset preparation to handle preprocessing manually.

> )
> 
> training_args.remove_unused_columns = False                         # Ensure unused columns are not removed in the collator (important for batch processing).
> 

Reviewer: ```suggestion
    remove_unused_columns=False,                         # Ensure unused columns are not removed in the collator (important for batch processing).
)
```

                if image is not None:
                    image = Image.open(io.BytesIO(image["bytes"]))
                    image_inputs.append(image.convert("RGB"))
    return image_inputs

def collate_fn(examples):
    texts = [processor.apply_chat_template(example["messages"], tokenize=False) for example in examples]
    if "images" in examples[0]:  # single-image
        images = [
            img.convert("RGB") if img.mode == "RGBA" else img for example in examples for img in example["images"]
        ]
    else:  # multi-image
        images = [process_vision_info(example["messages"]) for example in examples]

    # Tokenize the texts and process the images
    batch = processor(
        text=texts, images=images, return_tensors="pt", padding=True
    )  # Encode texts and images into tensors

    # The labels are the input_ids, and we mask the padding tokens in the loss computation
    labels = batch["input_ids"].clone()  # Clone input IDs for labels
    # Mask image tokens
    image_token_id = [
        processor.tokenizer.convert_tokens_to_ids(processor.tokenizer.special_tokens_map["boi_token"])
    ]
    # Mask tokens for not being used in the loss computation

>     labels[labels == processor.tokenizer.pad_token_id] = -100

Reviewer: Is the eos different from the pad token? because we we to mask the pad tokens, but not the eos tokens.

Author: It is 😄 

![Screenshot 2025-04-09 at 15 19 02](https://github.com/user-attachments/assets/c3a3d909-a0af-4252-a8e5-8c995c4fbf29)
